src/moaicore/MOAICCParticleSystem.cpp

```diff
@@ -0,0 +1,2389 @@
+			startIdx = this->mEndIndex;
+		}
+		
+		particle->mDeckIndex = (u32) Rand(startIdx, endIdx);
+	}
+	else { // sequential
+		u32 span = (this->mEndIndex - this->mStartIndex) + 1;
+		if (this->mEndIndex > this->mStartIndex) {
+			particle->mDeckIndex = this->mStartIndex + (this->mSeqIndex++ % span);
+		}
+		else{
+			particle->mDeckIndex = this->mStartIndex;
+		}
+	}
+	
+	
+}
+
+void MOAICCParticleSystem::InitializeEmitter () {
+	this->mEmissionRate = 1.0f / (this->mTotalParticles / this->mLifespan);
+	
+	float minLifespan = this->mLifespan - this->mLifespanVariance;
+	if ( minLifespan < 0.0f ) {
+		minLifespan = 0.0f;
+	}
+	
+	this->mLifespanTerm[0] = minLifespan;
+	
+	float maxLifespan = this->mLifespan + this->mLifespanVariance;
+	this->mLifespanTerm[1] = maxLifespan;
+}
+
+bool MOAICCParticleSystem::IsDone () {
+	return false;
+}
+
+
+bool MOAICCParticleSystem::IsFull () {
+	return (this->mParticleCount == this->mTotalParticles);
+}
+
+
+MOAICCParticleSystem::MOAICCParticleSystem() :
+	mParticles( NULL ),
+	mParticleCount( 0 ),
+	mAllocatedParticles( 0 ),
+	mTotalParticles( 0 ),
+	mEmitterType( EMITTER_GRAVITY ),
+	mLifespan( 0.0f ),
+	mLifespanVariance( 0.0f ),
+	mAngle( 0.0f ),
+	mAngleVariance( 0.0f ),
+	mStartSize( 16.0f ),
+	mStartSizeVariance( 0.0f ),
+	mFinishSize( 0.0f ),
+	mFinishSizeVariance( 0.0f ),
+	mMaxRadius( 0.0f ),
+	mMaxRadiusVariance( 0.0f ),
+	mMinRadius( 0.0f ),
+	mMinRadiusVariance( 0.0f ),
+	mRadialAcceleration( 0.0f ),
+	mRadialAccelVariance( 0.0f ),
+	mTangentialAcceleration( 0.0f ),
+	mTangentialAccelVariance( 0.0f ),
+	mRotStart( 0.0f ),
+	mRotStartVariance( 0.0f ),
+	mRotEnd( 0.0f ),
+	mRotEndVariance( 0.0f ),
+	mSpeed( 0.0f ),
+	mSpeedVariance( 0.0f ),
+	mRotPerSecond( 0.0f ),
+	mRotPerSecondVariance( 0.0f ),
+	mRotationalAcceleration( 0.0f ),
+	mRotationalAccelVariance( 0.0f ),
+	mDuration( -1.0f ),
+	mBlendFuncSrc( GL_ONE ),
+	mBlendFuncDst( GL_ONE_MINUS_SRC_ALPHA ),
+	mEmitCounter( 0.0f ),
+	mEmissionRate( 0.0f ),
+	mElapsed( 0.0f ),
+	mActive( false ),
+	mParticlePositionType( PARTICLE_POSITION_GROUPED ),
+	mIndexMode( INDEX_MODE_SEQUENTIAL ),
+	mStartIndex( 0 ),
+	mEndIndex( 0 ),
+	mDeckTilesX( 1 ),
+	mDeckTilesY( 1 ),
+	mDeckCellWidth( 1.0 ),
+	mDeckCellHeight( 1.0 ),
+	mDeckOffsetX( 0.0 ),
+	mDeckOffsetY( 0.0 ),
+	mDeckTileWidth( 1.0 ),
+	mDeckTileHeight( 1.0 ),
+	mSeqIndex( 0 )
+{
+	int i = 0;
+	for ( ; i < 2;  ++i) {
+		this->mLifespanTerm[i] = 0.0f;
+		this->mGravity[i] = 0.0f;
+		this->mGravityVariance[i] = 0.0f;
+		this->mSourcePos[i] = 0.0f;
+		this->mSourcePosVariance[i] = 0.0f;
+	}
+	
+	for (i = 0; i < 4; ++i) {
+		this->mStartColor[i] = 1.0f;
+		this->mStartColorVariance[i] = 0.0f;
+		this->mFinishColor[i] = 0.0f;
+		this->mFinishColorVariance[i] = 0.0f;
+	}
+	
+	RTTI_BEGIN
+		RTTI_EXTEND( MOAIProp )
+		RTTI_EXTEND( MOAIAction )
+	RTTI_END
+}
+
+MOAICCParticleSystem::~MOAICCParticleSystem () {
+	// clean up the allocated array
+	if (this->mParticles) {
+		free (this->mParticles);
```

Comment:
These should not be allocated using the C allocator. Bad Things (tm) will happen.